File: Meli/LeetCode 75/2130_Maximum_Twin_Sum_of_a_Linked_List.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        curr, prev = slow, None
        while curr:
            curr.next, prev, curr = prev, curr, curr.next
        
        answer = 0
        while prev:
            answer = max(answer, prev.val+head.val)
            prev = prev.next
            head = head.next

        return answer

File: Gombang/[LeetCode75] - [Linked List]/2130_Maximum_Twin_Sum_of_a_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        int maxSum = 0;
        vector<int> vec;

        while (head != nullptr)
        {
            vec.push_back(head->val);
            head = head->next;
        }

        for (int i = 0; i < vec.size() / 2; i++)
        {
            maxSum = max(maxSum, vec[i] + vec[vec.size() - i - 1]);
        }

        return maxSum;
    }
};